src/uuid.ts

```diff
@@ -1,57 +1,219 @@
-/**
- * UUID regular expression pattern copied from `uuid` npm module.
- * @see https://github.com/uuidjs/uuid/blob/master/src/regex.js
- */
-const UUID_RX = /^(?:[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}|00000000-0000-0000-0000-000000000000)$/i;
+import { Buffer } from 'buffer';
+import { v4 as uuidV4Generate } from 'uuid';
+import { ensureBuffer } from './ensure_buffer';
+import { Binary, BinaryExtended, BinaryExtendedLegacy } from './binary';
+import {
+  bufferToUuidHexString,
+  uuidHexStringToBuffer,
+  uuidHexStringValidateV4
+} from './uuid_utils';
+import type { EJSONOptions } from './extended_json';
 
 /** @public */
-export interface UUIDExtended {
+export type UUIDExtended = {
   $uuid: string;
-}
+};
+
+const BYTE_LENGTH = 16;
+
+const kId = Symbol('id');
 
 /**
- * Parser function copied from `uuid` npm module.
- * @see https://github.com/uuidjs/uuid/blob/master/src/parse.js
- * @internal
+ * A class representation of the BSON UUID type.
+ * @public
  */
-export function parseUUID(uuid: string): Uint8Array {
-  if (typeof uuid !== 'string') {
-    throw new TypeError('Invalid type for UUID, expected string but got ' + typeof uuid);
+export class UUID {
+  _bsontype = 'UUID' as const;
+
+  static cacheHexString: boolean;
+
+  /** UUID Bytes @internal */
+  private [kId]: Buffer;
+  /** UUID hexString cache @internal */
+  private __id?: string;
+
+  /**
+   * Create an UUID type
```

Comment:
`type` -> `object`

Reply:
Just staying with the theme: https://github.com/steffenagger/js-bson/blob/master/src/objectid.ts#L55
But you could very well argue that it should be `Create a UUID type` instead.